pkg/ingester/client/pool.go

```diff
@@ -1,78 +1,126 @@
 package client
 
 import (
+	"flag"
 	fmt "fmt"
 	io "io"
 	"sync"
 	"time"
 
 	"github.com/go-kit/kit/log/level"
 	"github.com/weaveworks/common/user"
+	"github.com/weaveworks/cortex/pkg/ring"
 	"github.com/weaveworks/cortex/pkg/util"
 	context "golang.org/x/net/context"
-	grpc_health_v1 "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/health/grpc_health_v1"
 )
 
-// Factory defines the signature for an ingester client factory
-type Factory func(addr string, cfg Config) (IngesterClient, error)
+// Factory defines the signature for an ingester client factory.
+type Factory func(addr string) (grpc_health_v1.HealthClient, error)
```

Comment:
Repeating this comment because GitHub hid it:
I think pushing this type into the public interface makes it clear the coupling between IngesterPool and gRPC is unhealthy.
How about defining an abstract interface containing some IsHealthy() method and push the gRPC details back to (in this case) IngesterClient ?

Reply:
I had a go at this; challenge is making the interface anything other the the same as the gRPC HealthClient interface makes using the pool harder (too hard IMO).  The coupling can be seen as intentional: this is a connection pool for `HealthClients`, adding the semantics of periodically removing unhealthy ones.  WDYT?